pymongo/uri_parser.py

```diff
@@ -53,7 +67,25 @@ def parse_userinfo(userinfo):
     # No password is expected with GSSAPI authentication.
     if not user:
         raise InvalidURI("The empty string is not valid username.")
-    return unquote_plus(user), unquote_plus(passwd)
+
+
+    # If we cannot round-trip a value, it is not properly URI encoded.
+    # This checks that we can take a value, quote it, and then unquote
+    # that quoted string and arrive at the same value.
+    # We take in the unquoted user and passwd strings and
+    # strip subdelimeters and then compare the original unquoted value
+    # without subdelimeters to the resulting string after round-tripping.
+    # One round trip means to quote once and then unquote.
+    for value_name, quoted_value in zip(
+            ('username', 'password',), (user, passwd,)):
+        quoted_value_no_sdelims = "".join(
+            [ch for ch in quoted_value if ch not in SUBDELIMS])
+        quoted_value = quote_plus(quoted_value_no_sdelims)
+        if quoted_value_no_sdelims != unquote_plus(quoted_value):
+            raise InvalidURI("%r is not a valid %s" % (
+                quoted_value, value_name))
```

Comment:
Please move this code up above `user, _, passwd = userinfo.partition(":")`. then we can remove the for-loop zip and instead validate once using `userinfo`.

Please also raise the same error as above `raise InvalidURI("Username and password must be escaped according to ...`